pygments/lexers/lean.py

```diff
@@ -122,5 +123,113 @@ class Lean3Lexer(RegexLexer):
         ],
     }
 
+    def analyse_text(text):
+        if re.search(r'^import [a-z]', text, re.MULTILINE):
+            return 0.1
+
 
 LeanLexer = Lean3Lexer
+
+
+class Lean4Lexer(RegexLexer):
+    """
+    For the Lean 4 theorem prover.
+    """
+    name = 'Lean4'
+    url = 'https://github.com/leanprover/lean4'
+    aliases = ['lean4']
+    filenames = ['*.lean']
+    mimetypes = ['text/x-lean4']
+    version_added = '2.18'
+
+    # same as Lean3Lexer, with `!` and `?` allowed
+    _name_segment = (
+        "(?![λΠΣ])[_a-zA-Zα-ωΑ-Ωϊ-ϻἀ-῾℀-⅏𝒜-𝖟]"
+        "(?:(?![λΠΣ])[_a-zA-Zα-ωΑ-Ωϊ-ϻἀ-῾℀-⅏𝒜-𝖟0-9'ⁿ-₉ₐ-ₜᵢ-ᵪ!?])*")
+    _name = _name_segment + r"(\." + _name_segment + r")*"
+
+    keywords1 = (
+        'import', 'abbreviation', 'opaque_hint', 'tactic_hint', 'definition',
+        'renaming', 'inline', 'hiding', 'parameter', 'lemma', 'variable',
+        'theorem', 'axiom', 'inductive', 'structure', 'universe', 'alias',
+        'help', 'options', 'precedence', 'postfix', 'prefix',
+        'infix', 'infixl', 'infixr', 'notation', '#eval',
+        '#check', '#reduce', '#exit', 'coercion', 'end', 'private', 'using', 'namespace',
```

Comment:
I know this list comes from the Lean 4 repo -- but just to note a few that I don't see used anywhere neither in core nor Mathlib:

* `coercion`
* `parameter`
* `help`
* `options`
* `including`
* `context`
* `expose`
* `print` / `reduce` / `synth` -- probably these are meant to be `#print`, `#reduce` / `#synth`?
* `opaque_hint` (looks like it was removed quite awhile ago in https://github.com/leanprover/lean4/commit/e3e2370a3806f30239643c4211ef51341113ba0f)
* `tactic_hint` (a quick search didn't find a specific commit removing it but it doesn't seem to exist anymore, all the commits referencing it are also from 2014)
* `reducible` / `irreducible` -- these only appear as attributes, not keywords -- I didn't double check whether they used to be (keywords) but if they're on this list nonetheless perhaps you should add `semireducible`


and a few that I encountered while double checking the above which *don't* appear here:

* `unif_hint` (appears in https://github.com/leanprover/lean4/blob/master/src/Init/Hints.lean#L11)
* `#check`
* `#check_failure`
* `#reduce`
* `#eval`
* `#synth`
* `#exit`
* `#print`
* `#print axioms`

Reply:
Done, I think. I also added `scoped` and `local`